### PR TITLE
Races between query completion and cancellation/timeout (JAVA-614, JAVA-632)

### PR DESCRIPTION
**Pull request by olim7t**
Includes a test for cancellation vs. successful response (first commit). I'm not entirely satisfied with it, so let's discuss it before extending it to other scenarios.

### CODE REVIEW

**Comment by olim7t** [comment]
Amended fix for 614 to handle edge case where `onSet` is called back immediately after `connection.write`. So we need to change the `queryState` before writing.

File: driver-core/src/test/java/com/datastax/driver/core/RequestHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import com.google.common.collect.ImmutableMap;
+import org.scassandra.Scassandra;
+import org.scassandra.http.client.PrimingRequest;
+import org.testng.annotations.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class RequestHandlerTest {
+
+    @Test(groups = "short")
```
Comment (tolbertam) [inline driver-core/src/test/java/com/datastax/driver/core/RequestHandlerTest.java]: This should probably be in the long group since it can take ~1 minute.

**Comment by olim7t** [comment]
Cleaned up history and force pushed.

**Comment by olim7t** [comment]
Merged to `2.0.9_fixes` branch. It doesn't show as merged here because this pull request was originally intended for 2.0.
